pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/SortedMailboxReceiveOperator.java

```diff
@@ -0,0 +1,189 @@
+    boolean foundNonNullTransferableBlock;
+    // For all non-singleton distribution, we poll from every instance to check mailbox content.
+    // Since this operator needs to wait for all incoming data before it can send the data to the
+    // upstream operators, this operator will keep trying to poll from all mailboxes until it only
+    // receives null blocks or EOS for all mailboxes. This operator does not need to yield itself
+    // for backpressure at the moment but once support is added for k-way merge when input data is
+    // sorted this operator will have to return some data blocks without waiting for all the data.
+    // TODO: Fix wasted CPU cycles on waiting for servers that are not supposed to give content.
+    do {
+      // Reset the following for each loop
+      foundNonNullTransferableBlock = false;
+      openMailboxCount = 0;
+      eosMailboxCount = 0;
+      for (int i = 0; i < _sendingMailbox.size(); i++) {
+        // this implements a round-robin mailbox iterator, so we don't starve any mailboxes
+        _serverIdx = (startingIdx + i) % _sendingMailbox.size();
+        MailboxIdentifier mailboxId = _sendingMailbox.get(_serverIdx);
+        try {
+          ReceivingMailbox<TransferableBlock> mailbox = _mailboxService.getReceivingMailbox(mailboxId);
+          if (!mailbox.isClosed()) {
+            openMailboxCount++;
+            TransferableBlock block = mailbox.receive();
+            // Get null block when pulling times out from mailbox.
+            if (block != null) {
+              foundNonNullTransferableBlock = true;
+              if (block.isErrorBlock()) {
+                cleanUpResourcesOnError();
+                _upstreamErrorBlock =
+                    TransferableBlockUtils.getErrorTransferableBlock(block.getDataBlock().getExceptions());
+                return _upstreamErrorBlock;
+              }
+              if (!block.isEndOfStreamBlock()) {
+                // Add rows to the PriorityQueue to order them
+                List<Object[]> container = block.getContainer();
+                _priorityQueue.addAll(container);
+              } else {
+                if (_opChainStats != null && !block.getResultMetadata().isEmpty()) {
+                  for (Map.Entry<String, OperatorStats> entry : block.getResultMetadata().entrySet()) {
+                    _opChainStats.getOperatorStatsMap().compute(entry.getKey(), (_key, _value) -> entry.getValue());
+                  }
+                }
+                eosMailboxCount++;
+              }
+            }
+          }
+        } catch (Exception e) {
+          return TransferableBlockUtils.getErrorTransferableBlock(
+              new RuntimeException(String.format("Error polling mailbox=%s", mailboxId), e));
+        }
+      }
+    } while (foundNonNullTransferableBlock && ((openMailboxCount > 0) && (openMailboxCount > eosMailboxCount)));
+
+    if (((openMailboxCount == 0) || (openMailboxCount <= eosMailboxCount))
```

Comment:
nit: I think changing the second condition to `openMailboxCount == eosMailboxCount` would be more readable, since openMailboxCount will never be less than eosMailboxCount.

Reply:
done